File: CS50/W1-C/credit/credit.c
#include <stdio.h>
#include <cs50.h>

// Functions used and defined below.
int get_num_digits(long number);
int get_first2digits(long number, int num_digits);
string card_bank(int first2digits, int num_digits);

int main(void)
{
    // Prompt user for credit card number.
    long card_number = get_long("Number: ");

    int second_last_x2 = 0;
    int step1_sum = 0;

    int step2_sum = 0;

    // Obtain the number of digits in the card number.
    int num_digits = get_num_digits(card_number);

    // Obtain the first two digits of the card number.
    int first2digits = get_first2digits(card_number, num_digits);

    /* Perform step 1 & step 2 of Luhn's Algorithm on every 2 digits in the
    card number starting from the right-most digits. */
    do
    {
        // Step 2, Part 1: From the last digit, add every other digit together.
        step2_sum += card_number % 10;

        // Remove last digit from card number since we're done with it.
        card_number /= 10;

        /* Step 1: From the 2nd to last digit, multiply every other digit by 2 and add the digit(s) of the result together. */
        second_last_x2 = (card_number % 10) * 2;
        if (second_last_x2 > 9)
        {
            step1_sum = (second_last_x2 / 10) + (second_last_x2 % 10);
        }
        else
        {
            step1_sum = second_last_x2;
        }

        // Step 2, Part 2: Add result of step 1 to step2_sum.
        step2_sum += step1_sum;

        // Remove 2nd last digit from card number since we're done with it.
        card_number /= 10;
    }
    while (card_number > 0);

    // Step 3 of Luhn's Algorithm: Check if last digit of step2_sum is 0.
    if (step2_sum % 10 == 0)
    {
        printf("%s", card_bank(first2digits, num_digits));
    }
    else
    {
        printf("INVALID\n");
    }
}
int get_num_digits(long number)
// Returns the number of digits contained in number.
{
    int i = 0;
    do
    {
        i += 1;
        number /= 10;
    }
    while (number > 0);
    return i;
}
int get_first2digits(long number, int num_digits)
// Returns the first two digits of number.
{
    for (int i = 0; i < num_digits - 2; i++)
    {
        number /= 10;
    }
    return number;
}
string card_bank(int first2digits, int num_digits)
/* Returns the bank for the credit card number or INVALID determined by using the first two digits and the number of digits in the card number. */
{
    if (num_digits == 15 && (first2digits == 34 || first2digits == 37))
    {
        return "AMEX\n";
    }
    if (num_digits == 16 && (first2digits >= 51 && first2digits <= 55))
    {
        return "MASTERCARD\n";
    }
    if ((num_digits == 13 || num_digits == 16) && first2digits / 10 == 4)
    {
        return "VISA\n";
    }
    return "INVALID\n";
}

File: CS50/W6-Python/sentimental-credit/credit.py
import re

from cs50 import get_string


def test_credit(number: str) -> None:
    """
    Test a credit card number and, if valid, prints the bank name.

    Prints INVALID if number is not validated as a credit card number.

    Args:
        number (str | int): The credit card number to test.
    """
    # Define dictionary containing tests for credit card name.
    basic_test = {'4': ({13, 16}, 'VISA'),
                  '5': ({16}, 'MASTERCARD'),
                  '3': ({15}, 'AMEX')}

    # Test to ensure the first digit(s) in the card are valid.  Based on the
    # following rules: Visa numbers start with 4; American Express numbers
    # start with 34 or 37; and MasterCard numbers start with 51, 52, 53, 54, or 55.
    pattern = re.compile(r'^(4.|3[47]|5[1-5])')
    match = pattern.match(number)
    if not match:
        print('INVALID')
        return None

    # Test if the number of digits align to first digit of the card number.
    # Based on the following rules: American Express uses 15-digit numbers;
    # MasterCard uses 16-digit numbers; and Visa uses 13- and 16-digit numbers
    num_digits = len(number)
    first_digit = match.group(1)[0]
    if num_digits not in basic_test[first_digit][0]:
        print('INVALID')
        return None

    # Perform steps 1 and 2 of Luhn’s Algorithm: (1) Starting from 2nd-to-last
    # digit, multiply every other digit by 2 and sum the digits of the results;
    # (2) Add step 1's sum to the sum of the remaining digits from the number
    # (the sum of every other digit from the last digit of the number).
    algorithm = (sum([int(i) * 2 // 10 + (int(i) * 2 % 10)
                      for i in number[-2::-2]])
                 + sum([int(i) for i in number[-1::-2]])
                 )

    # Perform step 3 of Luhn's Algorithm: Check if the sum from step 1 and 2 end with 0.
    if algorithm % 10 != 0:
        print('INVALID')
        return None

    # All tests passed. Print the name of the Financial Intitution for the card.
    print(basic_test[first_digit][1])


if __name__ == '__main__':
    card_number = get_string('Number: ')
    test_credit(card_number)
